server/src/main/java/io/crate/execution/dml/upsert/TransportShardUpsertAction.java

```diff
@@ -567,18 +554,14 @@ protected IndexItemResponse insert(Indexer indexer,
             indexer.updateTargets(addColumnResponse.addedColumns());
         }
 
-        if (indexer != null) {
-            // TODO: Add 2-phase indexing support to RawIndexer.
-            parsedDoc = indexer.index(item);
-        }
+        ParsedDocument parsedDoc = rawIndexer != null ? rawIndexer.index(item) : indexer.index(item);
 
         // Replica must use the same values for undeterministic defaults/generated columns
-        if (indexer.hasUndeterministicSynthetics()) {
-            item.insertValues(indexer.addGeneratedValues(item));
+        // This check must be done after index() call to let values/indexers size check compare original array sizes.
+        if (rawIndexer == null && indexer.hasUndeterministicSynthetics()) {
+            item.insertValues(indexer.addGeneratedValues(item.insertValues()));
```

Comment:
Initially I was fixing 
`test_generated_non_deterministic_value_is_consistent_on_primary_and_replica`

by doing same thing for RawIndexer if it's not null - but it's not related to the COPY FROM, issue was in the order of calls.

I suspect that master might have same issue but with COPY FROM - will verify and if it's really an issue, provide a separate PR against master.

Reply:
Actually I wonder if this call must not be forwarded to the `Indexer` used inside the `RawIndexer`? 🤔 
But as you also said, this is master related and related to your change.

Reply:
I pushed a small update, removed NPe check to sync with master.
https://github.com/crate/crate/pull/14696/commits/f7fee938c9f7c88ae520c79e04305026c9eff8b1.

I think on master it works since we mutate both refs/values and `RawIndexer` can see those changes.
With NPE check I had a wrong state (see screenshot). Not sure how it passed the test but I think it would be better to protect ourselves from having such state?

Reply:
<img width="1792" alt="Screenshot 2023-09-15 at 13 09 32" src="https://github.com/crate/crate/assets/7446940/3f29e54f-0f03-4161-bd27-8c8bfdb0bd7f">

Reply:
It was a bug on master, created https://github.com/crate/crate/pull/14702 to fix that